File: cracking_the_coding_interview/1-7.cpp
#include <iostream>
#include <vector>
using namespace std;

void print(const vector<vector<int>>& v){
    for(const vector<int>& line: v){
        for(int num: line){
            cout << num << " ";
        }
        cout << endl;
    }
    cout << endl;
}

void rotate(vector<vector<int>>& matrix){
    size_t size = matrix.size();
    for(int i=0; i < size / 2; i++){
        for(int j=0; j < size / 2; j++){
            int tmp = matrix[i][j];
            matrix[i][j] = matrix[size-j-1][i];
            matrix[size-j-1][i] = matrix[size-i-1][size-j-1];
            matrix[size-i-1][size-j-1] = matrix[j][size-i-1];
            matrix[j][size-i-1] = tmp;
        }
    }
    if(size % 2){
        int mid = size / 2;
        for(int i=0; i < size/2; i++){
            int tmp = matrix[i][mid];
            matrix[i][mid] = matrix[mid][i];
            matrix[mid][i] = matrix[size-i-1][mid];
            matrix[size-i-1][mid] = matrix[mid][size-i-1];
            matrix[mid][size-i-1] = tmp;
        }
    }
    return;
}

int main(void){
    vector<vector<int>> v1 = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}};
    print(v1);
    rotate(v1);
    print(v1);
    vector<vector<int>> v2;
    rotate(v2);
    print(v2);
    vector<vector<int>> v3 = {
        {1, 2, 3, 4},
        {5, 6, 7, 8},
        {9, 10, 11, 12},
        {13, 14, 15, 16}};
    print(v3);
    rotate(v3);
    print(v3);
    return 0;
}

File: cracking_the_coding_interview/1-7.py
import unittest

def rotate(matrix):
    N = len(matrix)
    for i in range((N + 1) // 2):
        for j in range(N // 2):
            tmp = matrix[i][j]
            matrix[i][j] = matrix[N - j - 1][i]
            matrix[N - j - 1][i] = matrix[N - i - 1][N - j - 1]
            matrix[N - i - 1][N - j - 1] = matrix[j][N - i - 1]
            matrix[j][N - i - 1] = tmp

class Test(unittest.TestCase):
    def test_even(self):
        matrix = [
            [0, 1, 2, 3, 4, 5],
            [6, 7, 8, 9, 10, 11],
            [12, 13, 14, 15, 16, 17],
            [18, 19, 20, 21, 22, 23],
            [24, 25, 26, 27, 28, 29],
            [30, 31, 32, 33, 34, 35]
        ]
        expected = [
            [30, 24, 18, 12, 6, 0],
            [31, 25, 19, 13, 7, 1],
            [32, 26, 20, 14, 8, 2],
            [33, 27, 21, 15, 9, 3],
            [34, 28, 22, 16, 10, 4],
            [35, 29, 23, 17, 11, 5]
        ]
        rotate(matrix)
        self.assertEquals(matrix, expected)

    def test_odd(self):
        matrix = [
            [0, 1, 2, 3, 4],
            [5, 6, 7, 8, 9],
            [10, 11, 12, 13, 14],
            [15, 16, 17, 18, 19],
            [20, 21, 22, 23, 24]
        ]
        expected = [
            [20, 15, 10, 5, 0],
            [21, 16, 11, 6, 1],
            [22, 17, 12, 7, 2],
            [23, 18, 13, 8, 3],
            [24, 19, 14, 9, 4]
        ]
        rotate(matrix)
        self.assertEquals(matrix, expected)

if __name__ == '__main__':
    unittest.main()